### PR TITLE
[Token] Add content hash

### PR DESCRIPTION
**Pull request by areshand**
### Description
motivation https://github.com/aptos-labs/aptos-core/issues/2939
### Test Plan

cargo test -p framework 

- - -
This change is [Reviewable](https://reviewable.io/reviews/aptos-labs/aptos-core/2991)

### CODE REVIEW

File: aptos-move/framework/aptos-token/sources/token.move [hunk 1]
```diff
@@ -96,6 +96,8 @@ module aptos_token::token {
         default_properties: PropertyMap,
         //control the TokenData field mutability
         mutability_config: TokenMutabilityConfig,
+        // content hash of the media
+        content_hash: String,
```
Comment (wrwg) [inline aptos-move/framework/aptos-token/sources/token.move:100]: Should we introduce a `Hash` type in the framework which has a 256 bit size? We discussed in some recent Move community meetings that independent of a native u256 type, such a hash type seems useful. You don't provide numeric algebra on hashes, so the set of operations is much smaller. We can use a relative bare bones Move `Hash` type for now, and then extend it over time. Basic point is whether we can agree on 256 bit precision.

It should probably be added to the `aptos_hash` module. The representation can be two `u128`, or whatever else division makes sense. This type would have just

**Comment by davidiw** [comment]
I'm concerned that this takes a very opinionated position about what a token is. This effectively says that it is an immutable piece of data off-chain. What exactly is this content and what does this hash represent? Is it the data at the URI? Is it the actual media content that is eventually pointed to? What if there is a lot of different pieces of data? What if the token author wants to mutate it over time, do they need to mutate this field first and then have a race condition on validity?

I think this belongs in the PropertyMap. But happy to chat more.

Also I know two other folks with strong opinions here: @CapCap and @aching

File: aptos-move/framework/aptos-token/sources/token.move (hunk 1 above)
Comment (areshand) [reply aptos-move/framework/aptos-token/sources/token.move:100]: sounds great. will try to modify the aptos_hash to introduce this `hash` type.

**Comment by areshand** [comment]
> I'm concerned that this takes a very opinionated position about what a token is. This effectively says that it is an immutable piece of data off-chain. What exactly is this content and what does this hash represent? Is it the data at the URI? Is it the actual media content that is eventually pointed to? What if there is a lot of different pieces of data? What if the token author wants to mutate it over time, do they need to mutate this field first and then have a race condition on validity?
> 
> I think this belongs in the PropertyMap. But happy to chat more.
> 
> Also I know two other folks with strong opinions here: @CapCap and @aching

Without content hash, it is hard to resolve the problems below:
1. the URI may not be permanently available or it will cost fee on creators.
2. the creator may not reveal the actually media files to public
3. the content of the uri can be updated. It is not quite sure which content the token claims ownership of. 

There are indeed many ways of generating the content hash. 

How to generate the content_hash? on what?
It is decided by the creator. It can be a content hash of image files or text, etc.
It can also be some other code that user can uniquely identify the claimed digital arts. 
The point is that they don't need to rely on URI or off-chain data to claim the ownership of a piece of content.
uri is more for ease of displaying the content. 

Mutate of content_hash?
no, this is like token name. we shouldn't allow user to mutate content_hash field. This is the token's semantic id. 
The creator should make sure the what ever content in their uri should match the content hash. 

Should this field be in property_map?
I don't have strong opinion here. I think it could be. I think we can make it optional. 

do content_hash solve the verification problem?
no, I think these days verifying a token is mainly checking if it is minted from THE "creator"'s account. The verification is verifying `who` created the token instead of `what content` this token claims. I think it works ok.

**Comment by areshand** [comment]
close this. This will be done in SDK

**Pull request closed without merging**